_posts/2024-02-09-Database-interaction-with-skunk-scala-library:-A-comprehensive-guide.md

```diff
@@ -0,0 +1,555 @@
+```
+User found: User(b68f2676-611e-4db2-af80-0458e9c52bd3,Jacob,jacob@email.com)
+```
+
+
+## 9. Refactoring our code
+Let's refactor our code in a functional style and add support for basic CRUD operations.
+Given we're leveraging the use of a purely functional library, we shall continue writing purely functional code throughout using the tagless final approach.
+
+### 9.1. Database Repository
+We define a trait called `Repository` to handle our queries and commands for interacting with our database. This file is placed in `src/main/scala/db`.
+```scala
+    import cats.effect.{Resource, Sync}
+    import cats.implicits.*
+    import skunk_guide.errors.Errors.UniqueViolation
+    import skunk.{Command, Query, Session, SqlState}
+
+    // While F is the effect type, E represents the domain/model used to perform operations
+    trait Repository[F[_], E](resource: Resource[F, Session[F]]) {
+      protected def run[A](session: Session[F] => F[A])(implicit F: Sync[F]): F[A] =
+        resource.use(session).handleErrorWith {
+          case SqlState.UniqueViolation(ex) =>
+            UniqueViolation(ex.detail.fold(ex.message)(m => m)).raiseError[F, A]
+        }
+
+      protected def findOneBy[A](query: Query[A, E], argument: A)(implicit F: Sync[F]): F[Option[E]] =
+        run { session =>
+          session.prepare(query).flatMap { preparedQuery =>
+            preparedQuery.option(argument)
+          }
+        }
+
+      protected def update[A](command: Command[A], argument: A)(implicit F: Sync[F]): F[Unit] =
+        run { session =>
+          session.prepare(command).flatMap { preparedCommand =>
+            preparedCommand.execute(argument).void
+          }
+        }
+    }
+```
+Our trait takes a `Resource` parameter useful to establish a database session. However, this is possible only for `Scala 3` as the app written in this article is based on Scala 3 features.
+
+We have a `run` method to allocate a resource to be used within a database session, a `findOneBy` method that can be used to fetch a specific row from a table such as fetching a user given the user's id, and an `update` method for operations that update the state of the database such as deleting from a database.
+`UniqueViolation` is just a postgres error exception which can be handled depending on the fields we set as unique in our tables.
+
+###  9.2. User Repository
+Still in the package `src/main/scala/db`, we define a `UserRepository` class to handle the `User` domain.
+```scala
+    import cats.effect.{Resource, Sync}
+    import cats.implicits.*
+    import skunk.{Codec, Command, Query, Session, Void, ~}
+    import skunk.codec.all.*
+    import skunk.implicits.*
+    import skunk_guide.domain.User
+    import skunk_guide.db.Repository
+    import fs2.Stream
+
+    import java.util.UUID
+
+    private final class UserRepository[F[_]: Sync](resource: Resource[F, Session[F]]) extends Repository[F, User](resource) {
```

Comment:
- Same comment on the Resource argument
- I'm not sure why make it private - who is this private for?